aws/resource_aws_wafv2_web_acl.go

```diff
@@ -0,0 +1,843 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/wafv2"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+)
+
+func resourceAwsWafv2WebACL() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsWafv2WebACLCreate,
+		Read:   resourceAwsWafv2WebACLRead,
+		Update: resourceAwsWafv2WebACLUpdate,
+		Delete: resourceAwsWafv2WebACLDelete,
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				idParts := strings.Split(d.Id(), "/")
+				if len(idParts) != 3 || idParts[0] == "" || idParts[1] == "" || idParts[2] == "" {
+					return nil, fmt.Errorf("Unexpected format of ID (%q), expected ID/NAME/SCOPE", d.Id())
+				}
+				id := idParts[0]
+				name := idParts[1]
+				scope := idParts[2]
+				d.SetId(id)
+				d.Set("name", name)
+				d.Set("scope", scope)
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"capacity": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"default_action": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"allow": wafv2EmptySchema(),
+						"block": wafv2EmptySchema(),
+					},
+				},
+			},
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringLenBetween(1, 256),
+			},
+			"lock_token": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(1, 128),
+					validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9-_]+$`), "must contain only alphanumeric hyphen and underscore characters"),
+				),
+			},
+			"scope": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					wafv2.ScopeCloudfront,
+					wafv2.ScopeRegional,
+				}, false),
+			},
+			"rule": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"action": {
```

Comment:
@anGie44  There are a few rules which I didn't exactly know how to validate when working on this, so only focused on basic implementation. I can look into it if you want.

But I don't think `default_action` and `action` are mutual exclusive 🤔

Reply:
I looked into some validation but think it's currently not possible because it's scoped on rules.

Would be nice to use `ValidateFunc` but I run into `ValidateFunc is not yet supported on lists or sets.`, so think there's not much possible currently :(

Reply:
thanks for the research @pvanbuijtene! i guess I had in mind using something like `ExactlyOneOf` but it does seem a bit restrictive in case things change at the AWS API level for some reason...the schema is most likely fine as is, I just wanted to bring it up for discussion as I saw the docs pointed out the use of 1 of the 2 fields but not both

Reply:
I also looked into `CustomizeDiff` but it looks like the the data for the `rules` is not available from the root, and the `CustomizeDiff` on the `rule` schema is not being called it looks like.

Reply:
noted, thank you! then let's keep things as they are here 👍